src/joystick/hidapi/SDL_hidapi_gip.c

```diff
@@ -1701,9 +1721,95 @@ static bool GIP_HandleCommandHidReport(
 {
-    // TODO
-    SDL_LogDebug(SDL_LOG_CATEGORY_INPUT, "GIP: Unimplemented HID Report message");
-    return false;
+    Uint64 timestamp = SDL_GetTicksNS();
+    // SDL doesn't have HID descriptor parsing, so we have to hardcode for the Chatpad descriptor instead.
+    // I don't know of any other devices that emit HID reports, so this should be safe.
+    if (attachment->attachment_type != GIP_TYPE_CHATPAD || !attachment->keyboard || num_bytes != 8) {
+        SDL_LogDebug(SDL_LOG_CATEGORY_INPUT, "GIP: Unimplemented HID Report message");
+        return false;
+    }
+
+    Uint8 modifiers = bytes[0];
+    Uint8 changed_modifiers = modifiers ^ attachment->last_modifiers;
+    if (changed_modifiers & 0x02) {
+        if (modifiers & 0x02) {
+            SDL_SendKeyboardKey(timestamp, attachment->keyboard, 0, SDL_SCANCODE_LSHIFT, true);
+        } else {
+            SDL_SendKeyboardKey(timestamp, attachment->keyboard, 0, SDL_SCANCODE_LSHIFT, false);
+        }
+    }
+    // The chatpad has several non-ASCII characters that it sends as Alt codes
+    if (changed_modifiers & 0x04) {
+        if (modifiers & 0x04) {
+            attachment->altcode_digit = 0;
+            attachment->altcode = 0;
+        } else {
+            if (attachment->altcode_digit == 4) {
+                char utf8[4] = {0};
+                // Some Alt codes don't match their Unicode codepoint for some reason
+                switch (attachment->altcode) {
+                case 128:
+                    SDL_UCS4ToUTF8(0x20AC, utf8);
+                    break;
+                case 138:
+                    SDL_UCS4ToUTF8(0x0160, utf8);
+                    break;
+                case 140:
+                    SDL_UCS4ToUTF8(0x0152, utf8);
+                    break;
+                case 154:
+                    SDL_UCS4ToUTF8(0x0161, utf8);
+                    break;
+                case 156:
+                    SDL_UCS4ToUTF8(0x0153, utf8);
+                    break;
+                default:
+                    SDL_UCS4ToUTF8(attachment->altcode, utf8);
+                    break;
+                }
+                SDL_SendKeyboardText(utf8);
+            }
+            attachment->altcode_digit = -1;
+            SDL_SendKeyboardKey(timestamp, attachment->keyboard, 0, SDL_SCANCODE_NUMLOCKCLEAR, true);
+            SDL_SendKeyboardKey(timestamp, attachment->keyboard, 0, SDL_SCANCODE_NUMLOCKCLEAR, false);
+        }
+    }
+
+    if (!bytes[2] && attachment->last_key) {
+        if (attachment->last_key == SDL_SCANCODE_CAPSLOCK) {
+            attachment->capslock = !attachment->capslock;
+        }
+        SDL_SendKeyboardKey(timestamp, attachment->keyboard, 0, attachment->last_key, false);
+        if (!(attachment->last_modifiers & 0xfd)) {
+            SDL_Keycode keycode = SDL_GetKeymapKeycode(NULL,
+                attachment->last_key,
+                (attachment->last_modifiers & 0x02) || attachment->capslock ? SDL_KMOD_SHIFT : 0);
```

Comment:
Parens for clarity?
```suggestion
                ((attachment->last_modifiers & 0x02) || attachment->capslock) ? SDL_KMOD_SHIFT : 0);
```

Reply:
I was waffling on if I should do this when I wrote the code. I can switch it.